File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        prefix = [-1] * n
        suffix = [n] * n 

        stack = []
        # suffix
        for i in range(n):
            while stack and heights[stack[-1]] > heights[i]:
                index = stack.pop()
                suffix[index] = i
            stack.append(i)
        # prefix
        stack = []
        for i in range(n - 1, -1, -1):
            while stack and heights[stack[-1]] > heights[i]:
                index = stack.pop()
                prefix[index] = i
            stack.append(i)

        maxArea = 0
        # calculation
        for i in range(n):
            width = suffix[i] - prefix[i] - 1
            height = heights[i]
            area = width * height
            maxArea = max(maxArea, area)
        return maxArea

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int[] newHeights = new int[n + 1];
        System.arraycopy(heights, 0, newHeights, 0, n);
        newHeights[n] = 0; 

        Deque<Integer> stack = new ArrayDeque<>();
        int maxArea = 0;

        for (int i = 0; i < newHeights.length; i++) {
            while (!stack.isEmpty() && newHeights[stack.peek()] > newHeights[i]) {
                int h = newHeights[stack.pop()];
                int w = stack.isEmpty() ? i : i - stack.peek() - 1;
                
                maxArea = Math.max(maxArea, h * w);
            }
            stack.push(i);
        }
        return maxArea;
    }
}